plugins/chat/assets/javascripts/discourse/services/chat-api.js

```diff
@@ -0,0 +1,248 @@
+import Service, { inject as service } from "@ember/service";
+import { ajax } from "discourse/lib/ajax";
+import UserChatChannelMembership from "discourse/plugins/chat/discourse/models/user-chat-channel-membership";
+import { tracked } from "@glimmer/tracking";
+import { bind } from "discourse-common/utils/decorators";
+import { Promise } from "rsvp";
+
+class Collection {
+  @tracked items = [];
+  @tracked meta = {};
+  @tracked loading = false;
+
+  constructor(resourceURL, handler) {
+    this._resourceURL = resourceURL;
+    this._handler = handler;
+    this._fetchedAll = false;
+  }
+
+  get loadMoreURL() {
+    return this.meta.load_more_url;
+  }
+
+  get totalRows() {
+    return this.meta.total_rows;
+  }
+
+  get length() {
+    return this.items.length;
+  }
+
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
+  [Symbol.iterator]() {
```

Comment:
It would be useful to add a comment here describing what this iterator does. The link to JS reference is helpful, but I can't figure out why we're iterating over the items here (and also not sure when the iteration takes place, on `init` I guess?).

Reply:
So Im still new to this, so not sure I can explain it better than reading the full mdn doc.

As far as I understand, by doing this we define the default iterator for an object. By defining the `next()` function we define the behavior of each iteration. As long as we have another value we return `{value, done: false}` and when we are at the end of iterables objects (items here) we return `{ done: true }`.

The iteration takes place when you use with forEach, for, map... It's very similar to ruby Enumerable where you define an `each` method.